File: ts/packages/core/src/source.ts
import {readFile} from "node:fs/promises";

export interface Source {
    getConfiguration: (environment: string, extension: string) => Promise<string> | string
}

export const FileSystemSource = (path: string) => ({
    getConfiguration: async (environment: string, extension: string) => {
        const dir = path.endsWith("/") ? path.slice(0, -1) : path
        return await readFile(`${dir}/${environment}.${extension}`, "utf8")
    }
} satisfies Source)

File: go/comfig/source.go
package comfig

import (
	"context"
	"fmt"
	"io/fs"
	"os"
)

type Source interface {
	Configuration(ctx context.Context, environment, extension string) ([]byte, error)
}

func NewFileSystemSource(fs fs.FS) Source {
	return fileSystemSource{fs: fs}
}

func NewFileSystemSourceByDirectory(dir string) Source {
	return NewFileSystemSource(os.DirFS(dir))
}

type fileSystemSource struct {
	fs fs.FS
}

func (s fileSystemSource) Configuration(_ context.Context, environment, extension string) ([]byte, error) {
	name := fmt.Sprintf("%s.%s", environment, extension)
	data, err := fs.ReadFile(s.fs, name)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", name, err)
	}
	return data, nil
}
